---
title: "Data viz with ggplot2"
author: "Vitor Aguiar"
date: "February 03, 2016"
output: 
  md_document:
  variant: github_markdown
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, comment='')
```

# Why visualize?

---

```{r}
anscombe
```

---

```{r}
# mean
sapply(anscombe, mean)

# variance
sapply(anscombe, var)

# correlation x y
mapply(cor, anscombe[1:4], anscombe[5:8])

# linear regression line
mapply(function(x, y) summary(lm(y ~ x))$coef[,1], anscombe[1:4], anscombe[5:8])
```

---

```{r, echo=FALSE}
library(ggplot2)
library(magrittr)
library(ggthemes)

anscombe2 <-
  anscombe %>% 
  dplyr::mutate(i = 1:nrow(.)) %>%
  tidyr::gather(variable, value, x1:y4) %>%
  tidyr::extract(variable, c("variable", "idx"), "(x|y)([1-4])") %>%
  tidyr::spread(variable, value)

ggplot(anscombe2, aes(x, y)) +
  geom_point() +
  geom_smooth(method = lm, se = FALSE) +
  facet_wrap(~idx, scales = "free") +
  labs(title = "Anscombe's Quartet of 'Identical' Simple Linear Regressions")
```

## Grammar of Graphics

gpplot2 is a system built on the **grammar of graphics** (Wilkinson, Anand and Grossman, 2005).

## Components:

- **data**
- **aes**: aesthetic mappings
- **geom**: geometric objects
- **scale**: control mapping from data to aesthetics
- **stat**: statistical transformation
- **coord**: coordinate system
- **facet**: faceting specification

## data

**Must be a data frame!**

- your data is very important, it's better to be clear
- a data frame is easier to save than a multitude of vectors: reproducibility

## aesthetic mappings

describe the way that variables are mapped to things that we can see on the plot

## geom

- control the type of plot you create (e.g., geom_point for scatterplots, geom_histogram for histograms)

## stat

- transforms the data, typically by summarising it in some manner

## position adjustments

- normally used with discrete data (bars)
- adjustments to deal with overlap

## scales

- control the mapping from data to aesthetics
- control all aspects of the guides (axes + legends)
- nomenclature: scale_aesthetic_nameofscale()
    - e.g.: scale_color_manual, scale_x_discrete, ...

---

see help(package="ggplot2")

## basic structure

```{r eval=FALSE}
ggplot(data, aes(x, y)) + geom
```

## The iris data

```{r}
str(iris)
```

# Difference between the base R plot system and ggplot2

# Scatterplot

---

```{r}
plot(iris$Sepal.Length, iris$Petal.Length)
```

---

```{r}
ggplot(iris, aes(Sepal.Length, Petal.Length)) + 
  geom_point()
```

# Multiple histograms 

## { .smaller}
```{r fig.width=5, fig.align='center'}
par(mfrow = c(3, 1))
hist(iris$Sepal.Length[iris$Species == "setosa"], col = "red", main = "setosa")
hist(iris$Sepal.Length[iris$Species == "versicolor"], col = "green", main = "versicolor")
hist(iris$Sepal.Length[iris$Species == "virginica"], col = "blue", main = "virginica")
```

## With ggplot2: { .smaller}

```{r}
ggplot(iris, aes(x = Sepal.Length, y = ..density.., fill = Species)) +
  geom_histogram(binwidth = 0.5) +
  facet_wrap(~Species, ncol = 1)
```

## automatic colors and legends

```{r}
ggplot(iris, aes(Sepal.Length, Petal.Length, color = Species)) +
  geom_point(size = 4)
```

## or shapes

```{r}
ggplot(iris, aes(Sepal.Length, Petal.Length, shape = Species)) +
  geom_point(size = 4)
```

## mapping size to a forth variable

```{r}
ggplot(iris, aes(Sepal.Length, Petal.Length, color = Species)) +
  geom_point(aes(size = Petal.Width))
```

## What if we wanted to color and separate plot windows by species, create a legend and add a smooth line?

---

```{r}
ggplot(iris, aes(Sepal.Length, Petal.Length, color = Species)) + 
  geom_point() +
  geom_smooth(method = lm) +
  facet_wrap(~Species, scales = "free_x")
```

# geoms

## smooth { .smaller}

```{r}
ggplot(iris, aes(Sepal.Length, Petal.Length)) +
  geom_point(size = 4) +
  geom_smooth(method = lm)
```

---

```{r}
ggplot(iris, aes(Sepal.Length, Petal.Length, color = Species)) +
  geom_point(size = 4) +
  geom_smooth(method = lm)
```

---

```{r}
ggplot(iris, aes(Sepal.Length, Petal.Length, color = Species)) +
  geom_point(size = 4) +
  geom_smooth(aes(group = 1), method = lm)
```

## jitter and boxplot

```{r}
ggplot(iris, aes(Species, Petal.Length, color = Species)) +
  geom_jitter()
```

---

```{r}
ggplot(iris, aes(Species, Petal.Length*Petal.Width, color = Species)) +
  geom_jitter()
```

---

```{r}
ggplot(iris, aes(Species, Petal.Length*Petal.Width, fill = Species)) +
  geom_boxplot()
```

## Histogram and density

```{r}
ggplot(iris, aes(Petal.Length)) + 
  geom_histogram(binwidth = .2) 
```

---

```{r}
ggplot(iris, aes(Petal.Length, fill = Species)) + 
  geom_histogram(binwidth = .1) 
```

---

```{r}
ggplot(iris, aes(Petal.Length, fill = Species)) + 
  geom_histogram(aes(y = ..density..), binwidth = .1) 
```

---

```{r}
# Density smooth
ggplot(iris, aes(Petal.Length, fill = Species)) + 
  geom_density(alpha = .7) +
  theme_bw()
```

## bar { .smaller}

```{r}
ggplot(msleep, aes(order, fill = vore)) + 
  geom_bar() +
  xlab("") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

# positioning in bar plots

---

```{r}
ggplot(dplyr::filter(msleep, !is.na(conservation)), 
       aes(vore, fill = conservation)) + 
  geom_bar(position = "stack")
```

---

```{r}
ggplot(dplyr::filter(msleep, !is.na(conservation)), 
       aes(vore, fill = conservation)) + 
  geom_bar(position = "fill")
```

---

```{r}
ggplot(dplyr::filter(msleep, !is.na(conservation)), 
       aes(vore, fill = conservation)) + 
  geom_bar(position = "dodge")
```

## text { .smaller}

```{r}
ggplot(msleep, aes(bodywt, sleep_total, label = genus, color = vore)) + 
  geom_text(size = 3) +
  scale_x_log10() +
  theme_bw() +  
  ylab("sleep (hours)")
```

# faceting

## { .smaller} 

- 1d: `facet_wrap()`
- 2d: `facet_grid()`

```{r}
ggplot(iris, aes(Petal.Length, fill = Species)) + 
  geom_density(color = NA) +
  facet_wrap(~Species, ncol = 1) + 
  theme_bw()
```

## more themes from the ggthemes package

```{r}
p <- 
  ggplot(iris, aes(Sepal.Length, Petal.Length, color = Species)) + 
  geom_point()
```

---

```{r}
p
```

##

```{r}
p + scale_color_wsj() + theme_wsj()
```

##

```{r}
p + scale_color_fivethirtyeight() + theme_fivethirtyeight()
```

##

```{r}
p + scale_color_economist() + theme_economist()
```

## lines { .smaller}

```{r fig.height=3.2, fig.width=6.4, fig.align='center'}
# allele frequency
f <- seq(0.05, 0.95, by = 0.05)

# genotype frequencies
hw <- 
  data.frame(allele_freq = f, f11 = f^2, f12 = 2*f*(1 - f), f22 = (1 - f)^2) %>%
  tidyr::gather(genotype, genotype_freq, f11:f22)

ggplot(data = hw, aes(x = allele_freq, y = genotype_freq, group = genotype)) +
  geom_line(size = 2)
```

## log transformation of the axes

```{r}
ggplot(data = msleep, aes(x = brainwt, y = bodywt, color = vore)) + 
  geom_point()
```

---

```{r}
ggplot(data = msleep, aes(x = log10(brainwt), y = log10(bodywt), 
                          color = vore)) + 
  geom_point()
```

---

```{r}
ggplot(data = msleep, aes(x = brainwt, y = bodywt, color = vore)) + 
  geom_point() +
  scale_x_log10() +
  scale_y_log10()
```

## Color

- hcl color scheme

- continuous:
    - `scale_color_gradient()` and `scale_fill_gradient()`
    - `scale_color_gradient2()` and `scale_fill_gradient2()`
    - `scale_color_gradientn()` and `scale_fill_gradientn()`
    
- discrete:
    - `scale_color_hue()` and `scale_fill_hue()`
    - `scale_color_brewer()` and `scale_fill_brewer()`
    - many schemes from `ggthemes` package


## RColorBrewer

```{r fig.height=7}
RColorBrewer::display.brewer.all()
```

## setting vs. mapping

```{r}
ggplot(data = iris, aes(x = Petal.Length, y = Petal.Width, 
                        color = "darkblue")) +
  geom_point()
```

---

```{r}
ggplot(data = iris, aes(x = Petal.Length, y = Petal.Width)) +
  geom_point(color = "darkblue")
```

---

```{r}
ggplot(data = msleep, aes(x = brainwt, y = bodywt, color = vore)) + 
  geom_point() +
  scale_x_log10() +
  scale_y_log10() +
  scale_colour_colorblind() #from ggthemes
```

# getting a bit more sophisticated

---

```{r echo=FALSE, fig.width=8}
source("../R/plot_funs.R")

weather_data <- 
  readr::read_csv("../data/dados_climaticos.csv") %>%
  tidyr::gather(variavel, valor, -(local:mes)) %>%
  tidyr::separate(variavel, c("variavel", "unidade"), 
                  sep = "_\\(|\\)_", extra = "drop") %>%
  dplyr::mutate(mes = factor(mes, levels = unique(mes)))

plot_weather(dplyr::filter(weather_data, variavel == "precipitação"))
```

---

```{r echo=FALSE, fig.width=9}
readr::read_csv("../data/dados_morfologicos.csv") %>%
dplyr::filter(variavel == "altura") %>%
plot_morpho(unit = "cm")
```

# Hands on