---
title: "reading and writing data"
author: "Vitor Aguiar"
date: "February 02, 2016"
output: ioslides_presentation
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, comment='')
```

## Base R

|   READ      |   WRITE     |
|:-----------:|:-----------:|
|read.table() |write.table()|
|read.csv()   |write.csv()  |
|read.csv2()  |write.csv2() |
|readLines()  |writeLines() |
|readRDS()    |saveRDS()    |

## package readr 

|   READ      |   WRITE     |
|:-----------:|:-----------:|
|read_table() |write_table()|
|read_csv()   |write_csv()  |
|read_csv2()  |write_csv2() |
|read_tsv()   |write_tsv()  |  
|read_rds()   |write_rds()  |

## package read_xl

- `excel_sheets()`
- `read_excel()`


## package data.table

for large data sets:

- `fread()` 

# Hands on

```{r eval=FALSE, include=FALSE}

## readxl
library(readxl)

morfo_excel <- "../data/dados_morfologicos.xlsx"

excel_sheets(morfo_excel)

altura_df <- read_excel(morfo_excel, na = "NA")

## base R

# write the altura data.frame to a file
# csv
write.csv(altura_df, "altura.csv")

# tab delimited
write.table(altura_df, "altura.txt", sep = "\t")

# space delimited
write.table(altura_df, "altura.txt")

# read it back
altura_df <- read.table("altura.txt")

# check the classes of each column with:
sapply(altura_df, class)

# some columns are factors because read.table has an argument stringsAsFactors=TRUE by default.

# illustrate the problem with factors

## readr
library(readr)

# write the altura data.frame to a file
write_csv(altura_df, "altura.csv")

# read it back
altura_df <- read_csv("altura.csv")

# check the classes of each column with:
sapply(altura_df, class)

# readr never converts strings into factors

# write the altura data.frame as a tab-delimited file
write_delim(altura_df, "altura.txt", delim = "\t")

# read directly from the web
# Gencode mouse gene annotation file
# this takes 1.7Gb of RAM on my Mac laptop
ftp_gtf <- "ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M7/gencode.vM7.annotation.gtf.gz"
mouse_annot <- read_delim(ftp_gtf, delim = "\t", comment = "##", col_names = FALSE) 

## data.table
library(data.table)

# csv
altura_df <- fread("altura.csv")

# tab
altura_df <- fread("altura.txt") 

# notice that fread() guesses the separator correctly
# check the class of altura_df
# objects of class data.table are manipulated differently
# for now, let's transform it into a data.frame.
# you can do that with the function as.data.frame() or dplyr's as_data_frame()

# vcf file
library(magrittr)

header <- 
  scan("~/mouse_500k_snps.vcf", what = "character", sep = "\n", nlines = 100) %>%
  .[grep("^##", .)]

mouse_vcf <- fread("~/mouse_500k_snps.vcf", header = TRUE, skip = length(header))


# try to read the files at:
http://ecologia.ib.usp.br/bie5782/doku.php?id=dados:start

## specialized tools

## the seqinr package

library(seqinr)

ftp_fasta <- "ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M7/gencode.vM7.transcripts.fa.gz"
download.file(ftp_fasta, destfile = "mouse_transcripts.fasta.gz")

mouse_seqs <- read.fasta("mouse_transcripts.fasta.gz", forceDNAtolower = FALSE,
                         as.string = TRUE, set.attributes = FALSE)
```